### PR TITLE
Core/Scripts: Implemented ScriptLoader Extensions

### PR DESCRIPTION
**Pull request by Seyden**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

The purpose of this PR is, to shorten the scripts.. we can outsource the header and the footer of scripts..
actually we're writing the header and footer in each script.

As example: https://gist.github.com/Seyden/8b4d849b2ad0b3bb481fe80637496398

- AuraScriptLoader extension
- SpellScriptLoader extension
- SpellAuraScriptLoader extension
- AreaTriggerEntityScriptLoader extension

**Target branch(es):** master

- [x] master

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)

- [x] tested ingame with a lot of scripts

**Known issues and TODO list:** (add/remove lines as needed)

No known issue

### CODE REVIEW

File: src/server/game/Scripting/ScriptMgr.h [hunk 1]
```diff
@@ -1171,6 +1171,55 @@ class TC_GAME_API ScriptMgr
         std::string _currentContext;
 };
 
+template <class _SpellScript>
+struct SpellScriptLoaderEx : public SpellScriptLoader
+{
+    SpellScriptLoaderEx(const char* name) : SpellScriptLoader(name) {}
+
+    SpellScript* GetSpellScript() const
```
Comment (joschiwald) [inline src/server/game/Scripting/ScriptMgr.h]: missing override

File: src/server/game/Scripting/ScriptMgr.h [hunk 2]
```diff
@@ -1171,6 +1171,55 @@ class TC_GAME_API ScriptMgr
         std::string _currentContext;
 };
 
+template <class _SpellScript>
+struct SpellScriptLoaderEx : public SpellScriptLoader
+{
+    SpellScriptLoaderEx(const char* name) : SpellScriptLoader(name) {}
```
Comment (joschiwald) [inline src/server/game/Scripting/ScriptMgr.h]: codestyle: const after type and one space between {}
`SpellScriptLoaderEx(char const* name) : SpellScriptLoader(name) { }`

File: src/server/game/Scripting/ScriptMgr.h [hunk 3]
```diff
@@ -1171,6 +1171,55 @@ class TC_GAME_API ScriptMgr
         std::string _currentContext;
 };
 
+template <class _SpellScript>
+struct SpellScriptLoaderEx : public SpellScriptLoader
```
Comment (Shauren) [inline src/server/game/Scripting/ScriptMgr.h]: SpellScriptLoader is a class so why are all of these struct?

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Scripting/ScriptMgr.h [hunk 4]
```diff
@@ -1171,6 +1171,55 @@ class TC_GAME_API ScriptMgr
         std::string _currentContext;
 };
 
+template <class _SpellScript>
+class SpellScriptLoaderEx : public SpellScriptLoader
+{
+    SpellScriptLoaderEx(char const* name) : SpellScriptLoader(name) { }
```
Comment (Shauren) [inline src/server/game/Scripting/ScriptMgr.h]: `public:`

**Comment by Seyden** [comment]
What is missing?

**Comment by tkrokli** [comment]
Approval from Shauren, I suppose.

**Comment by Vincent-Michael** [comment]
remove pls whitespaces

**Comment by Aokromes** [comment]
anything left here?

**Comment by tkrokli** [comment]
Approval from Shauren, maybe?

**Review by Shauren — APPROVED** [review]

**Comment by Shauren** [comment]
Code is ok, I don't like the `Ex` in name but at the same time I have no better idea

**Comment by Seyden** [comment]
@Shauren maybe there is a way to drop this redundant code completly? Isn't there a way to prevent using always SpellScript* GetSpellScript and this things?

**Comment by Seyden** [comment]
Don't merge it.. i will push a improvement tommorow

**Comment by Treeston** [comment]
@Shauren here's our new version after extensive IRC discussion earlier - I feel it's a lot cleaner than what was originally proposed.

PS: Advise on macro naming please, `REGISTER_SPELL_SCRIPT` (to keep with general style of allcaps for macros) vs `RegisterSpellScript` (matches with `PrepareSpellScript`)?

**Comment by Treeston** [comment]
Rebased for de-conflicting. Also includes example usage (the `spell_item` change).

**Pull request merged**